Use parent's index_name and document_type from child models

yuku-t opened the pull request:
Suppose there are three models:

``` rb
class Company < ActiveRecord::Base
  include Elasticsearch::Model
end
class TechCompany < Company; end
class TradingCompany < Company; end
```

then previously they respond to "document_type" as follows:

``` rb
Company.document_type        #=> 'company'
TechCompany.document_type    #=> 'tech_company'
TradingCompany.document_type #=> 'trading_company'
```

This commit changes them as:

``` rb
Company.document_type        #=> 'company'
TechCompany.document_type    #=> 'company'
TradingCompany.document_type #=> 'company'
```

So, we can use search scope like this:

``` rb
Company.search('*')  # Returns both TechCompany and TradingCompany

# Make `TechCompany.search` automatically add a filter such as
# `{ must: { term: { type: 'TechCompany' } } }`
TechCompany.search('*')  # Returns TechCompany only
```

yuku-t commented:
Sorry, I've just agreed to CLA.

karmi commented:
Interesting idea! Couple of notes, I'm travelling these days: we definitely need unit tests for this, and this is probably related to all the "single-table inheritance" (STI) issues being reported for Rails integration. Can you look at them? It would be great to have integration tests covering these...

karmi commented:
Hello, I'll need some time to process the patch -- there are couple of things which stick out, for instance, we deliberately don't add `active_record` as a dependency to the main .gemspec, but keep it in separate Gemfiles, etc.

yuku-t commented:
Hi, sorry for my late reply.

I miss understood the code base - I thought elasticsearch-model is used by only elasticsearch-rails. elasticsearch-persistence also uses it!

> we deliberately don't add `active_record` as a dependency to the main .gemspec, but keep it in separate Gemfiles

As I noted above, I added the dependency due to my wrong understanding. I've just removed it and updated the patch.

I want to write tests for this as soon as possible. (I'm busy now to write an article for Japanese famous tech magazine about "How to use elasticsearch in Ruby". It refers both this repository and elasticsearch-ruby :smirk: )

karmi commented:
@yuku-t Great news about the article! :) Take your time, there's no rush. Just ping me please when you have something ready, and ping me again if I get sidetracked and fail to respond.

BTW get in touch with @johtani for community support in Japan!

karmi commented:
Hi @yuku-t, did you had some time to return to this? Can you also please review #144 and #333?

yuku-t commented:
Hi, I finished the work last weekend. I check the failing tests and referred issues.

Tonkpils commented:
I'm assuming these two are related? #332

Tonkpils commented:
So, I tested out this changeset and the issue with this approach is that it ignores the `index_name` and `document_type` of the `base_class` if it's set. It will always use the model name instead of the proper base_class index/document.
